src/main/java/org/checkerframework/checker/noliteral/NoLiteralVisitor.java

```diff
@@ -0,0 +1,230 @@
+ * the bottom type as its default. Most of that code is either from or inspired by AWS' Data
+ * Classification Checker (https://github.com/awslabs/data-classification-checker).
+ */
+public class NoLiteralVisitor extends BaseTypeVisitor<NoLiteralAnnotatedTypeFactory> {
+  /**
+   * @param checker the type-checker associated with this visitor (for callbacks to {@link
+   *     TypeHierarchy#isSubtype})
+   */
+  public NoLiteralVisitor(BaseTypeChecker checker) {
+    super(checker);
+  }
+
+  /**
+   * Overrides the default lower bound for exception parameters. By default, this is top. This
+   * prevents the checker from defaulting exception parameters to @NonConstant. Overriding this
+   * method changes the expected lower bound to be @NonConstant, permitting that default.
+   *
+   * <p>Stolen shamelessly from https://github.com/awslabs/data-classification-checker
+   *
+   * @return a singleton set containing the @NonConstant annotation
+   */
+  @Override
+  protected Set<? extends AnnotationMirror> getExceptionParameterLowerBoundAnnotations() {
+    return Collections.singleton(atypeFactory.getCanonicalBottomAnnotation());
+  }
+
+  /**
+   * Overrides the default upper bound for thrown exceptions. By default, this is the same as the
+   * result of getExceptionParameterLowerBoundAnnotations, which defaults to top. This is overridden
+   * to keep the default top, since that other method is also overridden.
+   *
+   * <p>Stolen shamelessly from https://github.com/awslabs/data-classification-checker
+   *
+   * @return a singleton set containing the @MaybeConstant annotation
+   */
+  @Override
+  protected Set<? extends AnnotationMirror> getThrowUpperBoundAnnotations() {
+    return Collections.singleton(atypeFactory.getCanonicalTopAnnotation());
+  }
+
+  /**
+   * Searches through a type for user-written (i.e. non-default) annotations. Returns true if any
+   * are found. Stolen shamelessly from https://github.com/awslabs/data-classification-checker
+   *
+   * @param atm the type to search
+   * @return true if the type or any of its components has a non-default annotation
+   */
+  private boolean hasNonDefault(final AnnotatedTypeMirror atm) {
+    boolean result = false;
+    switch (atm.getKind()) {
+      case DECLARED:
+        AnnotatedTypeMirror.AnnotatedDeclaredType atmD =
+            (AnnotatedTypeMirror.AnnotatedDeclaredType) atm;
+        for (AnnotatedTypeMirror component : atmD.getTypeArguments()) {
+          if (hasNonDefault(component)) {
+            result = true;
+          }
+        }
+        break;
+      case ARRAY:
+        AnnotatedTypeMirror.AnnotatedArrayType atmA = (AnnotatedTypeMirror.AnnotatedArrayType) atm;
+        result = hasNonDefault(atmA.getComponentType());
+        break;
+      case WILDCARD:
+        return false;
+      case TYPEVAR:
+        return false;
+      default:
+        break;
+    }
+    return result || !atm.hasAnnotation(NonConstant.class);
```

Comment:
Do the base case first.  It is more efficient to check it first rather than after iterating through the type parameters and array levels.  I think it is also less confusing.